bliss/catalog.py

```diff
@@ -190,27 +196,44 @@ def to_full_catalog(self):
             pixel locations ("plocs") that are between 0 and `slen`.
         """
         # TODO: tile_slen should be an argument to this function, not stored in a tile catalog
-        plocs = self.get_full_locs_from_tiles()
-        param_names_to_mask = {"plocs"}.union(set(self.keys()))
-        tile_params_to_gather = {"plocs": plocs}
-        tile_params_to_gather.update(self)
+        if self.locations_sources:
+            plocs = self.get_full_locs_from_tiles()
+            param_names_to_mask = {"plocs"}.union(set(self.keys()))
+            tile_params_to_gather = {"plocs": plocs}
+            tile_params_to_gather.update(self)
+        else:
+            tile_params_to_gather = {}
+            tile_params_to_gather.update(self)
 
         params = {}
-        indices_to_retrieve, is_on_array = self.get_indices_of_on_sources()
+        if self.locations_sources:
+            indices_to_retrieve, is_on_array = self.get_indices_of_on_sources()
         for param_name, tile_param in tile_params_to_gather.items():
             if param_name == "n_sources":
                 continue
             if param_name == "locs":  # full catalog uses plocs instead of locs
                 continue
             k = tile_param.shape[-1]
+            # unsqueeze shear and conv
+            if param_name == "shear":  # pylint: disable=too-many-branches
+                continue
+            if param_name == "convergence":  # pylint: disable=too-many-branches
+                continue
             param = rearrange(tile_param, "b nth ntw s k -> b (nth ntw s) k", k=k)
-            indices_for_param = repeat(indices_to_retrieve, "b nth_ntw_s -> b nth_ntw_s k", k=k)
-            param = torch.gather(param, dim=1, index=indices_for_param)
-            if param_name in param_names_to_mask:
-                param = param * is_on_array.unsqueeze(-1)
+            if self.locations_sources:
+                indices_for_param = repeat(indices_to_retrieve, "b nth_ntw_s -> b nth_ntw_s k", k=k)
+                param = torch.gather(param, dim=1, index=indices_for_param)
+                if param_name in param_names_to_mask:
+                    param = param * is_on_array.unsqueeze(-1)
             params[param_name] = param
 
-        params["n_sources"] = reduce(self["n_sources"], "b nth ntw -> b", "sum")
+        if self.locations_sources:
+            params["n_sources"] = reduce(self["n_sources"], "b nth ntw -> b", "sum")
+
+        if "shear" in self:  # pylint: disable=too-many-branches
+            params["shear"] = self.get("shear")
+        if "convergence" in self:  # pylint: disable=too-many-branches
+            params["convergence"] = self.get("convergence")
```

Comment:
I'm reluctant to add special cases to this function that will make it even more complicated; it's so complex already. Is there a way to avoid these changes?

Reply:
this change was made since shear and convergence don't follow the generalized pattern requiring indices_of_on_sources  that the catalog mandates. will take a look at the version after your new PR and see what can be done here.

Reply:
resolved in refactor